### PR TITLE
ci: Enhanced code checking

### PR DESCRIPTION
**Pull request by hpohekar**
closes https://github.com/ansys/pyfluent/issues/1673

Integrated vulnerability check in pre-commit hooks which will run locally.

There is already [work in progress](https://github.com/ansys/actions/pull/666) for the deprecated command message, we will use it once it is ready. 

**Case 1 - check passed.**

```
$ git commit -m "ci: exit handling"
black................................................(no files to check)Skipped
isort................................................(no files to check)Skipped
flake8...............................................(no files to check)Skipped
codespell................................................................Passed
check yaml...............................................................Passed
pylint...............................................(no files to check)Skipped
Add License Headers..................................(no files to check)Skipped
Check Vulnerabilities Dependencies.......................................Passed
Check Vulnerabilities....................................................Passed
- hook id: check-vulnerabilities
- duration: 20.14s

+===========================================================================================================================================================================================+


DEPRECATED: this command (`check`) has been DEPRECATED, and will be unsupported beyond 01 June 2024.


We highly encourage switching to the new `scan` command which is easier to use, more powerful, and can be set up to mimic the deprecated command if required.


+===========================================================================================================================================================================================+




+===========================================================================================================================================================================================+


DEPRECATED: this command (`check`) has been DEPRECATED, and will be unsupported beyond 01 June 2024.


We highly encourage switching to the new `scan` command which is easier to use, more powerful, and can be set up to mimic the deprecated command if required.


+===========================================================================================================================================================================================+


[main]  INFO    profile include tests: None
[main]  INFO    profile exclude tests: B112,B403,B404,B110,B607,B314,B604,B311,B603,B307,B310,B405,B105,B602
[main]  INFO    cli include tests: None
[main]  INFO    cli exclude tests: None
Safety check performed.
Working... ---------------------------------------- 100% 0:00:10
[json]  INFO    JSON output written to file: info_bandit.json
Bandit check performed.
Advisory files generated successfully.
Dry run... not creating advisories and issues.
Information will be presented on screen.


*******************************************
Total 'safety' advisories detected: 0
Total 'safety' advisories reported: 0
Total 'bandit' advisories detected: 0
Total 'bandit' advisories reported: 0
*******************************************
Total advisories detected: 0
Total advisories reported: 0
*******************************************

[ci/enhanced_code_checking 7cd7c6ca3b] ci: exit handling
 1 file changed, 1 insertion(+), 1 deletion(-)
```

**Case 2 - check failed.**

```
$ pre-commit run --all-files
black....................................................................Passed
isort....................................................................Passed
flake8...................................................................Passed
codespell................................................................Passed
check yaml...............................................................Passed
pylint...................................................................Passed
Add License Headers......................................................Passed
Check Vulnerabilities Dependencies.......................................Passed
Check Vulnerabilities....................................................Failed
- hook id: check-vulnerabilities
- duration: 19.95s
- exit code: 1

+===========================================================================================================================================================================================+


DEPRECATED: this command (`check`) has been DEPRECATED, and will be unsupported beyond 01 June 2024.


We highly encourage switching to the new `scan` command which is easier to use, more powerful, and can be set up to mimic the deprecated command if required.


+===========================================================================================================================================================================================+




+===========================================================================================================================================================================================+


DEPRECATED: this command (`check`) has been DEPRECATED, and will be unsupported beyond 01 June 2024.


We highly encourage switching to the new `scan` command which is easier to use, more powerful, and can be set up to mimic the deprecated command if required.


+===========================================================================================================================================================================================+


[main]  INFO    profile include tests: None
[main]  INFO    profile exclude tests: B310,B604,B405,B602,B404,B110,B307,B112,B311,B403,B603,B314,B105
[main]  INFO    cli include tests: None
[main]  INFO    cli exclude tests: None
Safety check performed.
Working... ---------------------------------------- 100% 0:00:10
[json]  INFO    JSON output written to file: info_bandit.json
Bandit check performed.
Advisory files generated successfully.
Dry run... not creating advisories and issues.
Information will be presented on screen.

===========================================

Bandit [B607:start_process_with_partial_path] on ./src\ansys\fluent\core\fluent_connection.py - Hash: 03762c17822813383a8d90c4ece2ae0f

Starting a process with a partial executable path

#### Code

On file ./src\ansys\fluent\core\fluent_connection.py:

512     def _close_slurm(self):
513         subprocess.run(["scancel", f"{self._slurm_job_id}"])
514

#### CWE - 78
.
.
.
.

#### CWE - 78

For more information see https://cwe.mitre.org/data/definitions/78.html

#### More information

Visit https://bandit.readthedocs.io/en/1.8.2/plugins/b607_start_process_with_partial_path.html to find out more information.


*******************************************
Total 'safety' advisories detected: 0
Total 'safety' advisories reported: 0
Total 'bandit' advisories detected: 7
Total 'bandit' advisories reported: 7
*******************************************
Total advisories detected: 7
Total advisories reported: 7
*******************************************
```

### CODE REVIEW

**Comment by RobPasMue** [comment]
I would encourage your team to give it a local pass to the tool before enabling it on CI/CD. I'm sure Bandit will detect many issues and it will populate your advisories with a lot of draft ones... See https://actions.docs.ansys.com/version/stable/vulnerability-actions/index.html#check-vulnerabilities-action on how to run the tool locally

**Comment by hpohekar** [comment]
> I would encourage your team to give it a local pass to the tool before enabling it on CI/CD. I'm sure Bandit will detect many issues and it will populate your advisories with a lot of draft ones... See https://actions.docs.ansys.com/version/stable/vulnerability-actions/index.html#check-vulnerabilities-action on how to run the tool locally

@RobPasMue Sure. Thanks a lot.

**Review by prmukherj — APPROVED** [review]

**Review by RobPasMue — CHANGES_REQUESTED** [review]
Blocking PR until local resolutions have been performed by @hpohekar to avoid creation of unnecessary advisories

File: src/ansys/fluent/core/workflow.py [hunk 1]
```diff
@@ -298,8 +298,10 @@ def _task_by_id(task_id):
                         return mappings[task_id]
                     try:
                         return self._command_source._task_by_id(task_id)
-                    except Exception:
-                        pass
```
Comment (hpohekar) [inline src/ansys/fluent/core/workflow.py:302]: @prmukherj @seanpearsonuk @mkundu1 

try, except and pass is not allowed. We have used this block at many places in workflow.py. We should handle the error appropriately.

File: src/ansys/fluent/core/workflow.py [hunk 2]
```diff
@@ -162,10 +162,10 @@ def _call_refresh_task_accessors(obj):
     """This layer handles exception for PyConsole."""
     try:
         _refresh_task_accessors(obj)
-    except Exception:
-        # Is there a more specific Exception derived class
-        # for which we know it is correct to pass?
-        pass
+    except KeyError as e:
```
Comment (seanpearsonuk) [inline src/ansys/fluent/core/workflow.py:165]: @hpohekar Judging by the original comment, the exception is part of the regular workflow. This type of logical is plausible. As long as the comments are expenaded to explain what is going on in a bit more detail then it should have been fine as it was. Likewise, logging an **error** would be excessive.

File: src/ansys/fluent/core/workflow.py (hunk 1 above)
Comment (seanpearsonuk) [reply src/ansys/fluent/core/workflow.py:302]: > @prmukherj @seanpearsonuk @mkundu1
> 
> try, except and pass is not allowed. We have used this block at many places in workflow.py. We should handle the error appropriately.

Same comment applies here.

File: src/ansys/fluent/core/workflow.py (hunk 2 above)
Comment (seanpearsonuk) [reply src/ansys/fluent/core/workflow.py:165]: If we are sure that `KeyError` is completely accurate (backed up by testing) then retain that change. It's alo OK to remove the orignal comment since the question is answered by the exception type change.

Please change the logger call severity to debug since this is an example of EAFP (rather than an exceptional error) which is more normal in Python than most other languages.

File: src/ansys/fluent/core/workflow.py (hunk 1 above)
Comment (seanpearsonuk) [reply src/ansys/fluent/core/workflow.py:302]: @hpohekar Similar changes as above please.

**Comment by RobPasMue** [comment]
> @RobPasMue @seanpearsonuk
> 
> We are going to use this tool offline. We are not going integrate it into CI/CI workflow right now.

What's the reason for not integrating it into the workflow?

**Comment by hpohekar** [comment]
> > @RobPasMue @seanpearsonuk
> > We are going to use this tool offline. We are not going integrate it into CI/CI workflow right now.
> 
> What's the reason for not integrating it into the workflow?

@RobPasMue We want to integrate it with pre-commit hooks.

**Comment by RobPasMue** [comment]
> @RobPasMue We want to integrate it with pre-commit hooks.

That would be awesome. Once you have something, it might be a good idea to migrate it / make it part of https://github.com/ansys/pre-commit-hooks

**Comment by hpohekar** [comment]
> > @RobPasMue We want to integrate it with pre-commit hooks.
> 
> That would be awesome. Once you have something, it might be a good idea to migrate it / make it part of https://github.com/ansys/pre-commit-hooks

@RobPasMue Sure.

File: check_vulnernabilities.py
Comment (mkundu1) [inline check_vulnernabilities.py:1]: Is this taken from somewhere?

File: .ci/compare_flobject.py
Comment (mkundu1) [inline .ci/compare_flobject.py:1]: This file and the associated .yml can be deleted (in a future PR)

File: src/ansys/fluent/core/codegen/allapigen.py [hunk 3]
```diff
@@ -45,12 +45,15 @@ def generate(version: str, static_infos: dict, verbose: bool = False):
...
+        k: dict(v) for k, v in api_tree.items()
+    }  # Ensure all values are dicts for JSON serialization
     _update_first_level(api_tree, datamodelgen.generate(version, static_infos, verbose))
     _update_first_level(api_tree, settingsgen.generate(version, static_infos, verbose))
     api_tree_file = get_api_tree_file_name(version)
     Path(api_tree_file).parent.mkdir(parents=True, exist_ok=True)
-    with open(api_tree_file, "wb") as f:
-        pickle.dump(api_tree, f)
+    with open(api_tree_file, "w") as f:
+        json.dump(api_tree, f)
```
Comment (mkundu1) [inline src/ansys/fluent/core/codegen/allapigen.py:56]: I think we were no longer writing the pickle file. Some code can be cleaned up here (in a future PR).

File: src/ansys/fluent/core/codegen/tuigen.py [hunk 4]
```diff
@@ -35,17 +35,19 @@
...
 import shutil
 import string
-import subprocess
+import subprocess  # nosec B404
 from typing import Any, Dict
 import uuid
-import xml.etree.ElementTree as ET
+
+import defusedxml
+import defusedxml.ElementTree as ET
```
Comment (mkundu1) [inline src/ansys/fluent/core/codegen/tuigen.py:50]: Is this a new dependency?

File: src/ansys/fluent/core/codegen/tuigen.py [hunk 5]
```diff
@@ -295,12 +304,11 @@ def generate(self) -> None:
                         Path(__file__)
                         / ".."
                         / "data"
-                        / f"static_info_{self._version}_{self._mode}.pickle"
```
Comment (mkundu1) [inline src/ansys/fluent/core/codegen/tuigen.py:298]: This file is already checked in, we need to regenerate the equivalent .json file.

File: src/ansys/fluent/core/fluent_connection.py [hunk 6]
```diff
@@ -782,8 +782,8 @@ def _exit(
             if cleanup_on_exit:
                 try:
                     connection_interface.exit_server()
-                except Exception:
-                    pass
+                except Exception as e:
+                    logger.debug(f"Failed to exit server: {e}")
```
Comment (mkundu1) [inline src/ansys/fluent/core/fluent_connection.py:786]: I'm not aware of any scenario where server is not exited, the try-except is to catch any potential grpc error while returning from the remote call from an exiting server.

Maybe a better logging message is `Error during exiting server`.

File: src/ansys/fluent/core/utils/file_transfer_service.py [hunk 7]
```diff
@@ -264,7 +263,9 @@ def __init__(
         self.mount_target = mount_target if mount_target else "/home/container/workdir/"
         self.mount_source = mount_source if mount_source else MOUNT_SOURCE
         try:
-            self.host_port = port if port else random.randint(5000, 6000)
+            import secrets
```
Comment (mkundu1) [inline src/ansys/fluent/core/utils/file_transfer_service.py:266]: can be a global import?

File: tests/test_codegen.py [hunk 8]
```diff
@@ -61,12 +61,21 @@ def test_codegen_with_no_static_info(monkeypatch):
     version = "251"
     allapigen.generate(version, {})
     generated_paths = list(codegen_outdir.iterdir())
-    assert len(generated_paths) == 1
-    assert set(p.name for p in generated_paths) == {f"api_tree_{version}.pickle"}
```
Comment (mkundu1) [inline tests/test_codegen.py:65]: How the other asserts are fine, but not these?

File: src/ansys/fluent/core/get_build_details.py [hunk 9]
```diff
@@ -44,17 +46,25 @@ def get_build_version():
...
         build_details["Branch"] = (
-            subprocess.check_output(["git", "branch", "--show-current"])
+            subprocess.check_output(
+                ["git", "branch", "--show-current"]
+            )  # nosec B602 B603 B607
             .decode("ascii")
             .strip()
         )
-    except Exception:
-        pass
```
[the remark this answers is not in the document]
Comment (mkundu1) [reply src/ansys/fluent/core/get_build_details.py:57]: @hpohekar Please check if we can use https://bandit.readthedocs.io/en/latest/plugins/b110_try_except_pass.html to retain similar code as above.

**Comment by mkundu1** [comment]
@hpohekar For bandit, if there are some common patterns we are always allowing, can we specify them in a centralized configuration (like the pre-commit yaml) rather than polluting the source code?

File: src/ansys/fluent/core/utils/file_transfer_service.py [hunk 10]
```diff
@@ -433,8 +436,9 @@ def __init__(self, pim_instance: Any | None = None):
             try:
                 from simple_upload_server.client import Client
 
+                token = os.getenv("UPLOAD_SERVER_TOKEN", "token")
```
Comment (mkundu1) [inline src/ansys/fluent/core/utils/file_transfer_service.py:439]: Please document new env var in doc\source\contributing\environment_variables.rst

File: src/ansys/fluent/core/utils/file_transfer_service.py [hunk 11]
```diff
@@ -264,7 +263,9 @@ def __init__(
         self.mount_target = mount_target if mount_target else "/home/container/workdir/"
         self.mount_source = mount_source if mount_source else MOUNT_SOURCE
         try:
-            self.host_port = port if port else random.randint(5000, 6000)
+            import secrets
+
+            self.host_port = port if port else secrets.randbelow(1000) + 5000
```
Comment (mkundu1) [inline src/ansys/fluent/core/utils/file_transfer_service.py:268]: Issue in the original code: the deduced port can be unavailable in the host. A better logic here would be to run a loop between the lower and upper limit till an availabe port is found.

File: src/ansys/fluent/core/codegen/tuigen.py (hunk 4 above)
Comment (hpohekar) [reply src/ansys/fluent/core/codegen/tuigen.py:50]: No. We have removed it.

File: src/ansys/fluent/core/codegen/allapigen.py (hunk 3 above)
Comment (hpohekar) [reply src/ansys/fluent/core/codegen/allapigen.py:56]: Sure.

File: check_vulnernabilities.py
Comment (hpohekar) [reply check_vulnernabilities.py:1]: Yes, https://github.com/ansys/actions/tree/main/check-vulnerabilities. We need this to integrate it locally into pre-commit hooks.

File: src/ansys/fluent/core/codegen/allapigen.py (hunk 3 above)
Comment (hpohekar) [reply src/ansys/fluent/core/codegen/allapigen.py:56]: https://github.com/ansys/pyfluent/issues/3739

File: .ci/compare_flobject.py
Comment (hpohekar) [reply .ci/compare_flobject.py:1]: Sure. https://github.com/ansys/pyfluent/issues/3739

File: src/ansys/fluent/core/fluent_connection.py (hunk 6 above)
Comment (hpohekar) [reply src/ansys/fluent/core/fluent_connection.py:786]: Done. Thank you.

File: src/ansys/fluent/core/get_build_details.py (hunk 9 above)
[the remark this answers is not in the document]
Comment (hpohekar) [reply src/ansys/fluent/core/get_build_details.py:57]: Yes. Thank you.

File: src/ansys/fluent/core/utils/file_transfer_service.py (hunk 7 above)
Comment (hpohekar) [reply src/ansys/fluent/core/utils/file_transfer_service.py:266]: Done. Thank you.

**Comment by hpohekar** [comment]
> @hpohekar For bandit, if there are some common patterns we are always allowing, can we specify them in a centralized configuration (like in the pre-commit yaml)? That way we can avoid proliferation of `# nosec` in the source code?

Need to check.

File: src/ansys/fluent/core/utils/file_transfer_service.py (hunk 10 above)
Comment (hpohekar) [reply src/ansys/fluent/core/utils/file_transfer_service.py:436]: Done. Thank you.

File: src/ansys/fluent/core/codegen/tuigen.py (hunk 5 above)
Comment (mkundu1) [reply src/ansys/fluent/core/codegen/tuigen.py:298]: Pickle is unsafe when it deserializes untrusted data. As we have serialized the data here, we can retain the original code.

File: .pre-commit-config.yaml [hunk 12]
```diff
@@ -91,3 +91,22 @@ repos:
     - id: add-license-headers
       args:
       - --start_year=2021
+
+- repo: local
+  hooks:
+    - id: check-vulnerabilities-dependencies
```
Comment (RobPasMue) [inline .pre-commit-config.yaml:97]: I understand the need/interest to have it as part of the pre-commit stages but, in order to properly set up the tool and have it report them in case they appear, it is necessary that the check-vulnerabilities GitHub action is also set up

File: .ci/compare_flobject.py [hunk 13]
```diff
@@ -17,16 +17,16 @@ def compare_flobject():
     container_name = uuid.uuid4().hex
     is_linux = platform.system() == "Linux"
     subprocess.run(
-        f"docker container create --name {container_name} {image_name}",
+        [f"docker container create --name {container_name} {image_name}"],
```
[the remark this answers is not in the document]
Comment (hpohekar) [reply .ci/compare_flobject.py]: Yes, I agree. Even after converting command as list it was giving false positive warning so we have added this check to ignore list and reverted this change from everywhere. Thank you.

File: .pre-commit-config.yaml (hunk 12 above)
Comment (hpohekar) [reply .pre-commit-config.yaml:97]: @seanpearsonuk Please share your views. Thank you.

File: src/ansys/fluent/core/utils/file_transfer_service.py (hunk 11 above)
Comment (hpohekar) [reply src/ansys/fluent/core/utils/file_transfer_service.py]: Right, updated now. Thank you.

File: .pre-commit-config.yaml [hunk 14]
```diff
@@ -91,3 +91,33 @@ repos:
...
```
Comment (hpohekar) [inline .pre-commit-config.yaml:97]: This hook checks whether required dependencies are already installed or not. If not it installs them so that vulnerability check can be performed without dependency issue.

File: .pre-commit-config.yaml [hunk 15]
```diff
@@ -91,3 +91,33 @@ repos:
...
+      name: Check Vulnerabilities Dependencies
+      entry: python -c "import os; import subprocess; subprocess.run(['python', 'requirements/requirements_vulnernabilities.py'])"
+      language: python
+      types: [python]
+      pass_filenames: false
+      always_run: true
+
+    - id: check-vulnerabilities
```
Comment (hpohekar) [inline .pre-commit-config.yaml:105]: This hook performs actual vulnerability check and reports them if there are any. In the description of this PR we can see the actual reporting of vulnerabilities.

File: pyproject.toml [hunk 16]
```diff
@@ -218,3 +218,22 @@ showcontent = true
...
+    "B311",
+    "B314",
+    "B403",
+    "B404",
+    "B405",
+    "B602",
+    "B603",
+    "B604",
+    "B607",
+    ]
```
Comment (hpohekar) [inline pyproject.toml:239]: This makes whole source code `# nosec` free. We can expand it in future if needed.

**Comment by seanpearsonuk** [comment]
DEPENDENCY_CHECK_TOKEN = your GitHub PAT.

Any need to be concerned about this?

**Comment by hpohekar** [comment]
> DEPENDENCY_CHECK_TOKEN = your GitHub PAT.
> 
> Any need to be concerned about this?

@seanpearsonuk No. I have already discussed this with @RobPasMue.

**Comment by seanpearsonuk** [comment]
> > DEPENDENCY_CHECK_TOKEN = your GitHub PAT.
> > Any need to be concerned about this?
> 
> @seanpearsonuk No. I have already discussed this with @RobPasMue.

OK

**Comment by mkundu1** [comment]
> > > DEPENDENCY_CHECK_TOKEN = your GitHub PAT.
> > > Any need to be concerned about this?
> > 
> > 
> > @seanpearsonuk No. I have already discussed this with @RobPasMue.
> 
> OK

@hpohekar In check_vulnerabilities.py, the github integration seems to be used only for creatings issues or security advisories in the github repo. Maybe we can skip those for local run and remove the github integration?

A separate concern is to maintain a copy of a file from another repo in our repo, but we can address it later (may require some change in the other repo).

File: .pre-commit-config.yaml [hunk 17]
```diff
@@ -91,3 +91,33 @@ repos:
...
+      name: Check Vulnerabilities
+      entry: python -c 
+        "import os; 
+         import subprocess;
+         result = []; 
+         result.append(subprocess.run(['python', 'check_vulnerabilities.py', '--run-local'])) 
+         if os.environ.get('DEPENDENCY_CHECK_TOKEN') 
+         else print('DEPENDENCY_CHECK_TOKEN is not set');
```
Comment (seanpearsonuk) [inline .pre-commit-config.yaml:113]: @hpohekar We should skip the check without failure if the user hasn't set it up but also consider not having it in pre-commit. I favour the latter because it all seems heavyweight and risky for minimal payoff.

Apologies, I did not see your prior question here.

cc: @mkundu1 @RobPasMue

Comment (hpohekar) [reply .pre-commit-config.yaml]: @seanpearsonuk Removed.

**Comment by hpohekar** [comment]
> > > > DEPENDENCY_CHECK_TOKEN = your GitHub PAT.
> > > > Any need to be concerned about this?
> > > 
> > > 
> > > @seanpearsonuk No. I have already discussed this with @RobPasMue.
> > 
> > 
> > OK
> 
> @hpohekar In check_vulnerabilities.py, the github integration seems to be used only for creatings issues or security advisories in the github repo. Maybe we can skip those for local run and remove or avoid the github integration?
> 
> A separate concern is to maintain a copy of a file from another repo in our repo, but we can address it later (may require some change in the original file in the other repo).

@mkundu1 Done. Thank you.

**Review by mkundu1 — APPROVED** [review]

**Review by seanpearsonuk — APPROVED** [review]

**Comment by hpohekar** [comment]
> @mkundu1 @seanpearsonuk @hpohekar - as I posted here [#3731 (comment)](https://github.com/ansys/pyfluent/pull/3731#discussion_r1959176181), it is important that these checks are not only performed on pre-commit hooks but also on GitHub actions since they will be the ones reporting the vulnerabilities (if any) to our internal dashboard and on GitHub (privately, of course - users won't see them). Please address this request whenever you have the chance

@RobPasMue 

We have removed this from local pre-commit hook in [this PR](https://github.com/ansys/pyfluent/pull/3768). It was giving warnings for common Python packages as follows.

```
Safety vulnerability 64437 for package 'pillow'

Safety vulnerability 72132 for package 'zipp'
 
Safety vulnerability 73456 for package 'virtualenv'

Safety vulnerability 71608 for package 'urllib3'
```
